File: CombinationSumII.java
class Solution {
    List<List<Integer>> res;
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        List<Integer> cur = new ArrayList<>();
        res = new ArrayList<>();
        Arrays.sort(candidates);
        helper(candidates,target,cur,0,0);
        return res;
        
    }

    public void helper(int[] cand, int target,List<Integer> cur, int total, int i){
        if(total == target){
            res.add(new ArrayList<>(cur));
            return ;
        }

        if (i >= cand.length|| total > target){
            return;
        }

        for(int j = i; j <cand.length; j++){
            if(j >i && cand[j] == cand[j-1]) continue;
            cur.add(cand[j]);
            helper(cand,target,cur,total+cand[j],j+1);
            cur.remove(cur.size()-1);
        }

    }
}

File: Python/combinationSumII.py
class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        res = []
        def dfs(i,cur,total):
            if total == target:
                res.append(list(cur))
                return
            
            if i >= len(candidates) or total > target:
                return
            
            for j in range(i,len(candidates)):
                if (j > i and candidates[j] == candidates[j-1]):
                    continue
                cur.append(candidates[j])                
                dfs(j+1,cur,total+candidates[j])
                cur.pop()
        
        candidates.sort()
        dfs(0,[],0)
        return res